test/regression/2527_distortion_leak.cc

```diff
@@ -0,0 +1,173 @@
+/*
+ * Copyright (C) 2021 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+#include <mutex>
+
+#include "gazebo/physics/physics.hh"
+#include "gazebo/sensors/sensors.hh"
+#include "gazebo/common/common.hh"
+#include "gazebo/common/Timer.hh"
+#include "gazebo/rendering/Camera.hh"
+#include "gazebo/sensors/CameraSensor.hh"
+
+#include "gazebo/test/ServerFixture.hh"
+
+using namespace gazebo;
+
+class Issue2527Test : public ServerFixture
+{
+};
+
+std::mutex mutex;
+
+unsigned char* imgNormal = NULL;
+unsigned char* imgBarrel = NULL;
+unsigned char* imgPincushion = NULL;
+int imageCountNormal = 0;
+int imageCountBarrel = 0;
+int imageCountPincushion = 0;
+std::string pixelFormat = "";
+
+/////////////////////////////////////////////////
+void OnNewCameraFrame(int* _imageCounter, unsigned char* _imageDest,
+                  const unsigned char *_image,
+                  unsigned int _width, unsigned int _height,
+                  unsigned int _depth,
+                  const std::string &_format)
+{
+  std::lock_guard<std::mutex> lock(mutex);
+  pixelFormat = _format;
+  memcpy(_imageDest, _image, _width * _height * _depth);
+  *_imageCounter += 1;
+}
+
+/////////////////////////////////////////////////
+// \brief Test for issue #2527
```

Comment:
I believe you took the suggestion from https://github.com/osrf/gazebo/issues/2527#issuecomment-870123099 and have adapted the [CheckDistortion test from camera_sensor.cc](https://github.com/osrf/gazebo/blob/9cd47a0398da3c596e9dedc660cf1b1bb211118e/test/integration/camera_sensor.cc#L695). I think this is a good approach, and it would be worth adding a comment to that effect

Reply:
thanks for adding the comment in https://github.com/osrf/gazebo/pull/3044/commits/d0788d3a39910b41003aa83528b3254c3232d80e

I'm noticing that the camera parameters are different than the `CheckDistortion` test though. I'll review it a bit more closely to ensure the expectations make sense

Reply:
I just opened the test world and looked at the images, and the 12x12 resolution is too low in my opinion

![distortion_test_original](https://user-images.githubusercontent.com/3650755/126234278-c34151ca-d813-454a-befa-55f7d1e93f91.png)

I recompiled camera_sensor.cc with a `world->Save("/tmp/saved.world");` call after the camera sensors have been spawned in the `CheckDistortion` test and applied those changes to the test world in https://github.com/scpeters/gazebo/commit/c997c6bbb42bd4fa0036dc73fc56e86cf7317ed3 and was able to see the distortion more clearly:

![distortion_test_320x240](https://user-images.githubusercontent.com/3650755/126235745-a5a74551-a387-47c2-b002-ccb5cb399218.png)

I recommend changing the test world to use these parameters and updating the body of the test to more closely match the `CheckDistortion` test, such as matching the variable names where reasonable to make it easier to compare the tests in a tool like `meld`. You don't need to add the 4th camera that `CheckDistortion` has, I think 3 are enough

Reply:
Good catch! A 12x12 image is really impractical and would be too blocky to handle tests. I updated the camera resolution and updated the test code to match more similarly with the `CheckDistortion` test.